### ISSUE TITLE
Namelist file error checking

### ISSUE THREAD

**Issue by SeanBryan51** [CABLE-LSM/CABLE#431]
@aidanheerdegen recommended this technique for debugging bad namelist files: https://degenerateconic.com/namelist-error-checking.html

It would be great to have this for CABLE.

**Comment by aidanheerdegen** [comment]
@martindix suggested using `IOMSG`

> The UM (at least in CM2 and later versions) uses the IOMSG specifier introduced in Fortran 2003 which I think is the best approach. E.g.
> 
> ```fortran
> READ(UNIT=unit_in, NML=planet_constants, IOSTAT=ErrorStatus, IOMSG=iomessage)
> CALL check_iostat(errorstatus, "namelist planet_constants", iomessage)
>```
>
> Get error messages like
> ```
> invalid reference to variable in NAMELIST input, unit -129, file /scratch/p66/mrd599/my_namelist.nml, line 4, position 4
> ```

No reason you couldn't use both I guess.

**Comment by SeanBryan51** [comment]
Thanks - I didn't see the suggestion from Martin, I agree that would be the best approach

**Comment by Whyborn** [comment]
Yea, if we just pass the IO unit as an argument to the checker, we can get both. Having the description of the error and the line is really useful I think.

**Comment by Whyborn** [comment]
Related comment due to its appearance in the routine shown in https://degenerateconic.com/namelist-error-checking.html, should we adopt the ```iso_fortran_env``` module (see [here](https://fortranwiki.org/fortran/show/iso_fortran_env))?

**Comment by SeanBryan51** [comment]
> Related comment due to its appearance in the routine shown in https://degenerateconic.com/namelist-error-checking.html, should we adopt the iso_fortran_env module (see [here](https://fortranwiki.org/fortran/show/iso_fortran_env))?

I don't see how it is needed as the code doesn't appear to use `wp` (renamed to `real64`) anywhere in the code

**Comment by Whyborn** [comment]
I mean more generally- mainly for writing to STDOUT and STDERR, and specifying ```KIND``` for ```INTEGER``` and ```REAL``` variables.

**Comment by Whyborn** [comment]
I don't see why we wouldn't, I feel like we can safely say "too bad" to anyone using a compiler over 20 years old. I'd be happy at setting our standards at whatever was most recent 5 years ago, which at the moment would be Fortran 2018.

**Comment by SeanBryan51** [comment]
I agree. Other models are also moving to the ifx compiler. It would be good to have this information in the CABLE docs

**Comment by ccarouge** [comment]
Reading through the beginning of this discussion: https://fortran-lang.discourse.group/t/best-way-to-declare-a-double-precision-in-fortran/69/1  and the article linked to in one of the responses: https://stevelionel.com/drfortran/2017/03/27/doctor-fortran-in-it-takes-all-kinds/  would seem to indicate that using `SELECTED_xxx_KIND` intrinsics is the best way.

**Comment by Whyborn** [comment]
We've decided against including the method mentioned in the comic, as it relies on undefined behaviour which is subject to change at any time without notice (see [this thread](https://community.intel.com/t5/Intel-Fortran-Compiler/Regression-with-quot-backspace-quot-in-2023-0/m-p/1448175), and particular the response from Steve Lionel). We'll just use the ```IOMSG``` part.
